src/expint.jl

```diff
@@ -0,0 +1,331 @@
+import Polynomials
+using Base.MathConstants
+
+function E₁_cfpoly_approx(n::Integer, pstart::Polynomials.Polynomial{T}, ::Type{T}=BigInt) where {T<:Real}
+    q = Polynomials.Polynomial(T[1])
+    p = pstart
+    x = Polynomials.Polynomial(T[0,1])
+    for i = n:-1:1
+        p, q = x*p+(1+i)*q, p # from cf = x + (1+i)/cf = x + (1+i)*q/p
+        p, q = p + i*q, p     # from cf = 1 + i/cf = 1 + i*q/p
+    end
+    # do final 1/(x + inv(cf)) = 1/(x + q/p) = p/(x*p + q)
+    return p, x*p + q
+end
+
+macro E₁_cf64(x, n::Integer, start)
+    pstart = Polynomials.Polynomial(eval(start))
+    # consider using BigFloat?
+    p, q = E₁_cfpoly_approx(n, pstart, Float64)
+    xesc = esc(x)
+    
+    num_expr =  :(@evalpoly $xesc)
+    append!(num_expr.args, Float64.(Polynomials.coeffs(p)))
+    den_expr = :(@evalpoly $xesc)
+    append!(den_expr.args, Float64.(Polynomials.coeffs(q)))
+    :( exp(-$xesc) * $num_expr / $den_expr )
+end
+
+
+function E₁_taylor_coefficients(::Type{T}, n::Integer) where {T<:Number}
+    n < 0 && throw(ArgumentError("$n ≥ 0 is required"))
+    n == 0 && return T[]
+    n == 1 && return T[-eulergamma]
+    # iteratively compute the terms in the series, starting with k=1
+    term::T = 1
+    terms = T[-eulergamma, term]
+    for k=2:n
+        term = -term * (k-1) / (k * k)
+        push!(terms, term)
+    end
+    return terms
+end
+
+# inline the Taylor expansion for a given order n, in double precision
+macro E₁_taylor64(z, n::Integer)
+    c = E₁_taylor_coefficients(Float64, n)
+    zesc = esc(z)
+    taylor = :(@evalpoly $zesc)
+    append!(taylor.args, c)
+    :( $taylor - log($zesc) )
+end
+
+E₁(x::Real) = E₁(float(x))
+
+function E₁(x::Float64)
```

Comment:
IIRC correctly, this also works for `ComplexF64` values as long as `real(z) ≥ 0`; maybe that can be added back in, and it can fall back to the generic routine for `real(z) < 0`.

Reply:
For complex values, just replace `x < 3.0` etcetera with `x + absy < 3.0` where `x = real(z)` and `absy = abs(imag(z))`